internal/unionstore/art/art_iterator.go

```diff
@@ -318,197 +338,166 @@ func (it *baseIter) compare(other *baseIter) int {
+			it.nodes = append(it.nodes, *child)
 			it.idxes = append(it.idxes, inplaceIndex)
-			return it.next()
+			continue
 		}
+		it.nodes = it.nodes[:depth]
+		it.idxes = it.idxes[:depth]
+		if depth == 0 {
+			return nullArtNode
+		}
+		it.idxes[depth-1]++
 	}
-	it.nodes = it.nodes[:depth]
-	it.idxes = it.idxes[:depth]
-	if depth == 0 {
-		return nullArtNode
-	}
-	it.idxes[depth-1]++
-	return it.next()
 }
 
 func (it *baseIter) prev() artNode {
-	depth := len(it.nodes) - 1
-	curr := it.nodes[depth]
-	idx := it.idxes[depth]
-	idx--
-	switch curr.kind {
-	case typeNode4:
-		n4 := it.allocator.getNode4(curr.addr)
-		idx = min(idx, int(n4.nodeNum)-1)
-		if idx >= 0 {
-			it.idxes[depth] = idx
-			child := n4.children[idx]
-			if child.kind == typeLeaf {
-				return child
+	for {
+		depth := len(it.nodes) - 1
+		curr := it.nodes[depth]
+		idx := it.idxes[depth]
+		idx--
+		var child *artNode
+		switch curr.kind {
+		case typeNode4:
+			n4 := it.allocator.getNode4(curr.addr)
+			idx = min(idx, int(n4.nodeNum)-1)
+			if idx >= 0 {
+				it.idxes[depth] = idx
+				child = &n4.children[idx]
+			} else if idx == inplaceIndex {
+				it.idxes[depth] = idx
+				if !n4.inplaceLeaf.addr.IsNull() {
+					return n4.inplaceLeaf
+				}
 			}
-			it.nodes = append(it.nodes, child)
-			it.idxes = append(it.idxes, node256cap)
-			return it.prev()
-		} else if idx == inplaceIndex {
-			it.idxes[depth] = idx
-			if !n4.inplaceLeaf.addr.IsNull() {
-				return n4.inplaceLeaf
+		case typeNode16:
+			n16 := it.allocator.getNode16(curr.addr)
+			idx = min(idx, int(n16.nodeNum)-1)
+			if idx >= 0 {
+				it.idxes[depth] = idx
+				child = &n16.children[idx]
+			} else if idx == inplaceIndex {
+				it.idxes[depth] = idx
+				if !n16.inplaceLeaf.addr.IsNull() {
+					return n16.inplaceLeaf
+				}
 			}
-		}
-	case typeNode16:
-		n16 := it.allocator.getNode16(curr.addr)
-		idx = min(idx, int(n16.nodeNum)-1)
-		if idx >= 0 {
-			it.idxes[depth] = idx
-			child := n16.children[idx]
-			if child.kind == typeLeaf {
-				return child
+		case typeNode48:
+			n48 := it.allocator.getNode48(curr.addr)
+			if idx >= 0 {
+				idx = n48.prevPresentIdx(idx)
 			}
-			it.nodes = append(it.nodes, child)
-			it.idxes = append(it.idxes, node256cap)
-			return it.prev()
-		} else if idx == inplaceIndex {
-			it.idxes[depth] = idx
-			if !n16.inplaceLeaf.addr.IsNull() {
-				return n16.inplaceLeaf
+			if idx >= 0 {
+				it.idxes[depth] = idx
+				child = &n48.children[n48.keys[idx]]
+			} else if idx == inplaceIndex {
+				it.idxes[depth] = idx
+				if !n48.inplaceLeaf.addr.IsNull() {
+					return n48.inplaceLeaf
+				}
 			}
-		}
-	case typeNode48:
-		n48 := it.allocator.getNode48(curr.addr)
-		if idx >= 0 {
-			idx = n48.prevPresentIdx(idx)
-		}
-		if idx >= 0 {
-			it.idxes[depth] = idx
-			child := n48.children[n48.keys[idx]]
-			if child.kind == typeLeaf {
-				return child
+		case typeNode256:
+			n256 := it.allocator.getNode256(curr.addr)
+			if idx >= 0 {
```

Comment:
Ditto, or add some comments to explain why the else branch does not need to be checked.